src/Storages/MergeTree/MergeTreeIndexTextPostingListCodec.cpp

```diff
@@ -0,0 +1,358 @@
+    {
+        current.clear();
+        total_in_current_segment = 0;
+        prev_value = {};
+    }
+
+    /// Flush current segment:
+    /// - compress pending partial block
+    /// - reset block state so a new segment can start
+    void flushSegment()
+    {
+       chassert(total_in_current_segment <= posting_list_block_size);
+       if (!current.empty())
+           compressBlock(current);
+
+        reset();
+    }
+
+    /// Write all segments to output and fill TokenPostingsInfo:
+    /// - offsets: byte offsets in output where each segment begins
+    /// - ranges: [min,max] docID range for each segment
+    void serializeTo(WriteBuffer & out, TokenPostingsInfo & info) const
+    {
+        for (const auto & desc : segments)
+        {
+            info.offsets.emplace_back(out.count());
+            info.ranges.emplace_back(desc.min, desc.max);
+            ContainerHeader header(desc.compressed_data_size, desc.count, desc.min);
+            header.encodeTo(out);
+            out.write(compressed_data.data() + desc.compressed_data_offset, desc.compressed_data_size);
+        }
+    }
+    /// Compress one block of delta values and append it to `compressed_data`.
+    ///
+    /// Block layout:
+    ///   [1 byte bits][payload]
+    ///
+    /// - bits: max bit-width among deltas in this block
+    /// - payload: Codec::encode(...) bitpacked bytes
+    ///
+    /// Also updates current segment metadata (count, max, payload size).
+    void compressBlock(std::span<uint32_t> segment)
+    {
+        segments.back().count += segment.size();
+        segments.back().max = prev_value;
+
+        auto [cap, bits] = BlockCodec::evaluateSizeAndMaxBits(segment.data(), segment.size());
+        size_t memory_gap = compressed_data.capacity() - compressed_data.size();
+        if (cap + 1 > memory_gap)
+        {
+            auto min_need = cap + 1 - memory_gap;
+            compressed_data.reserve(compressed_data.size() + 2 * min_need);
+        }
+        /// Block Layout: [1byte(bits)][payload]
+        size_t offset = compressed_data.size();
+        compressed_data.resize(compressed_data.size() + cap + 1);
+        auto * p = reinterpret_cast<unsigned char *>(compressed_data.data() + offset);
+
+        encodeU8(bits, p);
+        auto used = BlockCodec::encode(segment.data(), segment.size(), bits, p);
+        chassert(used == cap);
+
+        segments.back().compressed_data_size = compressed_data.size() - segments.back().compressed_data_offset;
+        current.clear();
+    }
+
+    /// Decode one compressed block into `current` and reconstruct absolute docIDs.
+    ///
+    /// - Reads bits-width byte
+    /// - Codec::decode fills `current` with delta values
+    /// - inclusive_scan converts deltas -> docIDs using `prev_value` as initial prefix
+    /// - Updates prev_value to the last decoded docID
+    static void decodeOneBlock(unsigned char * & in, size_t count, uint32_t & prev_value, std::vector<uint32_t> & current)
```

Comment:
In previous codes where we tried FastPFor compression, delta scan like similar to this used to impact performance significantly for big posting lists. We must keep an eye on this.